### ISSUE TITLE
Check why 0.4.x requires an older boost on Mac

### ISSUE THREAD

**Issue by axic** [ethereum/solidity#6349]
See https://github.com/ethereum/homebrew-ethereum/pull/221#discussion_r266425192

Also optionally fix this issue with 0.4.26.

**Comment by axic** [comment]
Was this fixed or just removed from 0.4.26?

**Comment by chriseth** [comment]
I do not remember.

**Comment by cameel** [comment]
Here's something related: tests on solc versions < 0.6.0 require boost < 1.71.0 to pass. That's because boost 1.71.0 fixed `boost::multiprecision` to respect `std::uppercase` when written to a stream as a hexadecimal number (https://github.com/boostorg/multiprecision/pull/130, https://github.com/boostorg/multiprecision/issues/129). Earlier boost versions were hard-coded to print hex numbers in uppercase.

Some of our tests relied on that bug - e.g. in solc 0.5.0 test case `Assembler/all_assembly_items` fails . In solc 0.5.13 most of our code was changed to use `std::uppercase` (#7567) but the one affecting that test case seems to have only been included in solc 0.6.0 (#7619, not sure why, maybe it was only included in `breaking`?).

**Comment by cameel** [comment]
As for solc 0.4.x, I needed boost older than 1.70.0 to build solc 0.4.0 on MacOS 10.15 (Github Actions) because of this error:
```c++
/Users/runner/work/solc-bin/solc-bin/solidity/libevmasm/Assembly.cpp:223:74: error: no matching conversion for functional-style cast from 'const dev::u256' (aka 'const number<boost::multiprecision::cpp_int_backend<256, 256, boost::multiprecision::unsigned_magnitude, boost::multiprecision::unchecked, void> >') to 'std::__1::string' (aka 'basic_string<char, char_traits<char>, allocator<char> >')
                                        createJsonValue("PUSH [tag]", i.location().start, i.location().end, string(i.data())));
                                                                                                            ^~~~~~~~~~~~~~~
```
`i.data()` returns `u256` (i.e. a specialization of `boost::multiprecision`) and apparently it's convertible to `string` only in older boost versions. For me specifically boost 1.63.0 worked for this specific error (though I'm still getting other, unrelated errors). I haven't tried any versions in between yet.

**Comment by cameel** [comment]
I'm pretty sure that the problem to investigate here (not being able to build 0.4.x with boost 1.69.0 and having to downgrade it to 1.60.0) is what I described in my comment above.
- boost 1.70.0 fails due to the problem with `multiprecision` described above.
- boost 1.65.0 works with all solidity versions between 0.3.6 and 0.6.8. Most likely with newer too.
- For many solidity versions boost 1.63.0 works too, though I don't remember if I tried it specifically with 0.4.26. It worked with at least some 0.4.x versions. I switched to boost 1.65.0 at some point.
- I was not able to try anything below that. boost 1.60.0 cannot be installed directly from homebrew core any more (`Error: boost@1.60 has been disabled because it is a versioned formula!`) and when checking out the formula from an earlier revision it fails too because it depended on something that has been moved/removed. I was only able to build versions 1.63.0 and up.

Also note that solc versions below 0.5.0 were not using boost unit test properly which results in `duplicate symbol` errors when linking (#3845). It was fixed in #4572 and the fix can be easily backported to earlier versions. I think affects all boost versions though.

@chriseth @axic is there anything more to investigate or should we close this issue?
